// gera/build/extensions/cli-extension.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
Object.defineProperty(exports, "__esModule", { value: true });
// add your CLI-specific functionality here, which will then be accessible
// to your commands
module.exports = function (toolbox) {
    toolbox.gera = {
        getCommands: function () {
            var commands = [];
            var arquivos = toolbox.filesystem.list(toolbox.filesystem.cwd() + '/.gera/');
            if (arquivos === undefined)
                return commands;
            var arquivoFilter = arquivos.filter(function (e) { return e.indexOf('.js') > 0; });
            arquivoFilter.forEach(function (arquivo) {
                var command = require(toolbox.filesystem.cwd() + '/.gera/' + arquivo);
                var typeofName = typeof command.name;
                var typeorRun = typeof command.run;
                if (typeofName !== 'string') {
                    toolbox.print.error("Erro module '" + arquivo + "', did not export the attribute string name (ex. module.exports  = {name:'<name_module>',...}).");
                }
                if (typeorRun !== 'function') {
                    toolbox.print.error("Erro module '" + arquivo + "', did not export the function run (ex. module.exports  = {run:<function_run>,...}).");
                }
                if (typeorRun === 'function' && typeofName === 'string') {
                    commands.push(command);
                }
            });
            return commands;
        },
        diretoryTemplates: function () {
            return toolbox.filesystem.cwd() + '/.gera/templates';
        },
        generate: function (props) { return __awaiter(void 0, void 0, void 0, function () {
            var template;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        template = toolbox.filesystem.cwd() + '/.gera/templates' + '/' + props.template;
                        if (toolbox.filesystem.exists(template) !== 'file') {
                            toolbox.print.error("template not found. Target: " + template);
                            return [2 /*return*/];
                        }
                        return [4 /*yield*/, toolbox.template.generate({
                                template: props.template,
                                target: props.target,
                                props: props.props,
                                directory: toolbox.gera.diretoryTemplates()
                            })];
                    case 1:
                        _a.sent();
                        return [2 /*return*/];
                }
            });
        }); }
    };
    // enable this if you want to read configuration in from
    // the current folder's package.json (in a "gera" property),
    // gera.config.json, etc.
    // toolbox.config = {
    //   ...toolbox.config,
    //   ...toolbox.config.loadConfig("gera", process.cwd())
    // }
};
